### ISSUE TITLE
Retry transactions failed for insufficient gas

### ISSUE THREAD

**Issue by stefanopepe** [near/near-wallet#2111]
## Summary

With the increased adoption of NEAR Protocol, we may experience peaks of usage that translate into higher gas costs. Some dapps attach a fixed amount of gas to a transaction, taking the risk of failing the transfer and put the user in an endless loop of "failed transactions" if no proper mitigations are in place.

This feature wants to unblock the user who is signing a transaction with too low gas, allowing them to retry manually attaching a larger amount. 

## Desired Outcome

As a user, I want to "bump" the gas fees on my failed transaction, such that I can make sure that I completed the request.

As a dapp owner, I want to make sure that my users can anyway use my platform if the wallet (or the network) misallocate their gas fees.

Success will be measured against the number of transactions failed for insufficient gas, and the number of transactions submitted right after that were successful.

## How will it work?

NEAR Wallet will submit the transaction as requested, and check if it failed, returning as usual the reason. If the failure reason is "exceeded prepaid gas", the UI will allow the user to retry the same transaction attaching more gas.
 
A first release of the transaction may leave to the user the decision on how much more to attach, but ideally it should suggest a number based on the most recent "retries".

This feature will support dapp developers who will be able to see how many users manually "bumped" transactions, allowing them to improve their application behavior.

**Comment by corwinharrell** [comment]
@stefanopepe @MaximusHaximus I've encountered a fundamental issue in designing around this that I wanted to surface and get your feedback on:

We currently don't calculate the cost of a transaction before it is executed. Right now, few simply display an arbitrary number labelled "estimated fees", which hopes to cover the cost of the TXN. As long as this is the case, we have no way to telling the user _how much_ gas might be required to ensure a successful TXN. If we are allowing the user to manually indicate the amount of gas they wish to use for a "bumped" TXN, they are taking a stab in the dark.

How should we handle this? Do we just allow them to guess? This is poor UX. IMO, a dependency of this work is calculating the actual TXN cost immediately before execution, and then offering the user a most realistic estimate for an effective increase.

**Comment by stefanopepe** [comment]
@corwinharrell I see the issue. Mine is a high-level proposal that didn't take into account this specific limitation.

I'm happy with keeping the scope of this feature as a simple "retry transaction" functionality that will implicitly attach more fees (e.g. bump to 300Tgas) and try again if the transaction didn't go through for insufficient gas.

If I understand correctly, this solution would preserve the desired outcome, keeping any complicated gas estimation&management on hold.

**Comment by corwinharrell** [comment]
@stefanopepe the only issue with this is that the user has no control over the amount of gas used as an increase. I guess since it is likely such a small amount that the change to their balance will almost be imperceptible to them, but it's worth considering. How should new determine the amount used as an increase? Do we need to reference a gas figure in the UI in addition to the cost in NEAR / Yocto?

**Comment by stefanopepe** [comment]
With @MaximusHaximus we were thinking of simply doubling the fees.
The reason is that any overpayment is given back to the user, so let's say that a transaction is appending 200Tgas but it's insufficient and should be 250, we append 400Tgas and the user gets back 150 (OVERSIMPLIFYING).
So the UI Would say "Attach more fees, no worries any overpayment will come back to your wallet, see _here_"

The only potential issue with this approach is that users need more available balance to effectively use this feature, but it's really unlikely given the upcoming update to sharding and the legwork around fees prediction.

**Comment by corwinharrell** [comment]
<img width="500" alt="Screen Shot 2021-09-30 at 3 58 58 PM" src="https://user-images.githubusercontent.com/5934188/135522295-34708db9-023b-497e-8395-7a6641be4ee5.png">

@stefanopepe @MaximusHaximus Curious to know if you think this is oversimplified or sufficient for a v1?

**Comment by marcinbodnar** [comment]
@corwinharrell Please send a Figma link for it, or text only.

**Comment by corwinharrell** [comment]
@marcinbodnar just waiting on feedback before I share the final design deliverables

**Comment by stefanopepe** [comment]
Looks good to me! My only concern is (for old-school crypto users) if the failed transaction was 0.001 before, why it should show the same amount (0.001) also when it asks to confirm again. As you know, Bitcoin and Ethereum offer to "bump the fees", showing a higher price in the native currency.
The issue is tricky to represent in NEAR tokens, as we need to increase the Tgas attached to the transaction - which is a separate unit of account and it is not related to the 0.001 needed to pay for it.

I don't see a major issue, but an opportunity to link users to the part of the documentation that explains "why the transaction cost seems the same to me?"

**Comment by corwinharrell** [comment]
@stefanopepe to clarify, can we not have the price increase reflected in the estimated fees? Just trying to understand why our initial estimated fees are calculated in NEAR, but the increase is calculated in Tgas.

**Comment by stefanopepe** [comment]
Absolutely yes, for a normal user is sufficient to know that the system will submit the same transaction with additional fees.
It's necessary to support power users as well (thinking of Ref.Finance tribe), considering a button to see additional transaction details and what will be done under the trunk (including the amount of Tgas attached) >> see #2164

**Comment by corwinharrell** [comment]
@marcinbodnar The design for this specific addition can be found here as part of the transaction signature updates that @Patrick1904 is working on: https://www.figma.com/file/Jl4i7oFfkM7uMxBidkZ3Tk/Signature-Screens?node-id=380%3A497

<img width="1182" alt="Screen Shot 2021-10-12 at 10 24 09 AM" src="https://user-images.githubusercontent.com/5934188/136974585-f0937a1c-69a9-4a8d-b5ca-8a5286cab008.png">

Should the transaction fail due to an insufficient fee limit, the transaction content will be replaced with this component allowing the user to resubmit the transaction with an increased fee limit. Should the user opt to cancel, the transaction would just be cancelled entirely, and they would need to start from scratch.

**Comment by stefanopepe** [comment]
I followed with interest the PR #2216 and I want to show an interesting corner case:

[this transaction](https://explorer.near.org/transactions/AvqjubrefbXk3f6DSNzQBarX7KgV5BkQDUYypkApPHJa) failed because it attached too much, 300Tgas, to the multisignature request.
The source of the failure is the original request, nested inside a 2FA `add_request` json, [attaching 300Tgas](https://explorer.near.org/transactions/ESqEQtACfarfL3V1kYCoEUyWewfbcKwGy28wLU2W86Zk) to it:
```
{
  "request": {
    "receiver_id": "dc293ffb9ac4c90199411a29f7326c8f06e49798.lockup.near",
    "actions": [
      {
        "type": "FunctionCall",
        "method_name": "unstake_all",
        "args": "e30=",
        "deposit": "0",
        "gas": "300000000000000"
      }
    ]
  }
}
```

The very same [nested transaction](https://explorer.near.org/transactions/ACLCAEqFRAc8aj9dR4Ar7gojm9jeSXbZ1QGvG9jhqCmF), attaching 250Tgas to the unstake method, [succeeded](https://explorer.near.org/transactions/Cv1wSAbPwLSd3ezY8hokQK2aFKY1n3k3YQz3ggHSuK9F):
```
{
  "request": {
    "receiver_id": "dc293ffb9ac4c90199411a29f7326c8f06e49798.lockup.near",
    "actions": [
      {
        "type": "FunctionCall",
        "method_name": "unstake_all",
        "args": "e30=",
        "deposit": "0",
        "gas": "250000000000000"
      }
    ]
  }
}
```

We can't change this piece of information, since it's stored inside the 2FA state, so we may need to just fail the transaction without trying any gas whatsoever OR introduce way more complex logic in presence of 2FA transactions.
